pkg/watch/watch.go

```diff
@@ -60,45 +77,70 @@ type InstrumentedListerWatcher struct {
 	metrics           *ListWatchMetrics
 	resource          string
 	useAPIServerCache bool
+	limit             int64
 }
 
 // NewInstrumentedListerWatcher returns a new InstrumentedListerWatcher.
-func NewInstrumentedListerWatcher(lw cache.ListerWatcher, metrics *ListWatchMetrics, resource string, useAPIServerCache bool) cache.ListerWatcher {
+func NewInstrumentedListerWatcher(lw cache.ListerWatcher, metrics *ListWatchMetrics, resource string, useAPIServerCache bool, limit int64) cache.ListerWatcher {
 	return &InstrumentedListerWatcher{
 		lw:                lw,
 		metrics:           metrics,
 		resource:          resource,
 		useAPIServerCache: useAPIServerCache,
+		limit:             limit,
 	}
 }
 
 // List is a wrapper func around the cache.ListerWatcher.List func. It increases the success/error
 // / counters based on the outcome of the List operation it instruments.
+// It supports setting object limits, this means if it is set it will only list and process
+// n objects of the same resource type.
 func (i *InstrumentedListerWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
 
 	if i.useAPIServerCache {
 		options.ResourceVersion = "0"
 	}
 
+	if i.limit != 0 {
+		options.Limit = i.limit
+		i.metrics.ListObjectsLimit.WithLabelValues(i.resource).Set(float64(i.limit))
+	}
+
 	res, err := i.lw.List(options)
+
 	if err != nil {
-		i.metrics.ListTotal.WithLabelValues("error", i.resource).Inc()
+		i.metrics.ListRequestsTotal.WithLabelValues("error", i.resource).Inc()
 		return nil, err
 	}
 
-	i.metrics.ListTotal.WithLabelValues("success", i.resource).Inc()
+	list, err := meta.ExtractList(res)
+	if err != nil {
+		return nil, err
+	}
+	i.metrics.ListRequestsTotal.WithLabelValues("success", i.resource).Inc()
+
+	if i.limit != 0 {
```

Comment:
We'd want to validate against negative values, also s/int/uint preferably. I'm reviewing this on my back from work so apologies if I missed something obvious here.

Reply:
Good point, to be honest, I have no idea why the kubernetes API made limits a signed int. https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ListOptions